sympy/solvers/solveset.py

```diff
@@ -1332,6 +1338,83 @@ def _is_logarithmic(f, symbol):
     return rv
 
 
+def _is_bivariate(f, symbol):
+    r"""
+    Helper to check if the equation takes bivariate form.
+    If the equation has exactly two generators then we can
+    say it to be in bivariate form.
+    """
+    poly = f.as_poly()
+    gens = _filtered_gens(poly, symbol)
+
+    return len(gens) == 2
+
+
+def _solve_as_bivariate(lhs, rhs, symbol, domain):
+    r"""
+    Helper solver to solve equations of bivariate form.
+    """
+    result = ConditionSet(symbol, Eq(lhs - rhs, 0), domain)
+
+    poly = lhs.as_poly()
+    gens = _filtered_gens(poly, symbol)
+    gpu = bivariate_type(lhs - rhs, *gens)
+    if gpu is not None:
+        g, p, u = gpu
+        if g != lhs:
+            usol = _solveset(p, u, domain)
+            if not isinstance(usol, ConditionSet):
+                result = [_solveset(g - us, symbol, domain) for us in usol]
+    return result
+
+
+def _is_lambert(f, symbol):
+    r"""
+    Helper to check whether an equation is of lambert type.
+    Equations containing `Pow`, `log` or `exp` terms are currently checked
+    for having lambert solutions
+    """
+    return any(isinstance(arg, (Pow, exp, log)) for arg in _term_factors(f))
+
+
+def _compute_lambert_solutions(lhs, rhs, symbol):
+    """
+    Computes the lambert solutions. Returns `None` if it
+    fails doing so.
+    """
+    try:
+        poly = lhs.as_poly()
+        gens = _filtered_gens(poly, symbol)
+
+        return _solve_lambert(lhs - rhs, symbol, gens)
+    except NotImplementedError:
+        pass
+
+
+def _solve_as_lambert(lhs, rhs, symbol, domain):
+    r"""
+    Helper solver to handle equations having LambertW solutions.
+    First tries to solve equation directly, if unsuccessful
+    attempts to find the solutions by making the `symbol` positive.
+    The second attempt modifies the `domain` having only posiive values.
```

Comment:
It's not clear what the significance of the domain being modified is. Does this mean a solution will be missed?

typo: posi*t*ive

Reply:
The idea of modifying the domain is simply because we will try to find solution by making `x` (variable to be solved) as positive, now since `x` is now reduced to positive from real, I belive the domain of the accepted values should be reduced to only positive.
I haven't come around a case where solution is being missed. Even if there comes such a case, shouldn't we include only positive values (considering we have assumed `x` to be positive)?

Reply:
No. This is anti-solveset. `solveset` must return all the solutions or the conditions under which the solutions are constrained. Making a symbol positive might help us find a solution, but solveset has to be introspective enough to know if it can't solve something completely. So finding a solution through symmetry is nice (e.g. that x = b in x**b - b**x is a solution) but if we don't know how to solve that equation in general then we have to return something indicating that.

Reply:
Otherwise solveset will just be a fancier version of solve...

Reply:
Alright, I got the point, it's better not to reduce the domain.